Common/src/main/java/com/blamejared/crafttweaker/api/tag/manager/type/KnownTagManager.java

```diff
@@ -0,0 +1,175 @@
+import com.blamejared.crafttweaker.api.util.GenericUtil;
+import com.blamejared.crafttweaker.mixin.common.access.tag.AccessTag;
+import com.blamejared.crafttweaker_annotations.annotations.Document;
+import com.mojang.datafixers.util.Pair;
+import net.minecraft.core.Holder;
+import net.minecraft.core.Registry;
+import net.minecraft.resources.ResourceKey;
+import net.minecraft.resources.ResourceLocation;
+import net.minecraft.tags.Tag;
+import net.minecraft.tags.TagManager;
+import org.openzen.zencode.java.ZenCodeType;
+
+import javax.annotation.Nullable;
+import java.util.ArrayList;
+import java.util.Collections;
+import java.util.HashMap;
+import java.util.List;
+import java.util.Map;
+import java.util.Optional;
+import java.util.stream.Collectors;
+
+@ZenRegister
+@Document("vanilla/api/tag/manager/type/KnownTagManager")
+@ZenCodeType.Name("crafttweaker.api.tag.manager.type.KnownTagManager")
+public class KnownTagManager<T> implements ITagManager<KnownTag<T>>, Iterable<KnownTag<T>> {
+    
+    private final ResourceKey<? extends Registry<T>> resourceKey;
+    private final Class<T> elementClass;
+    private final MutableLoadResult<T> backingResult;
+    private Map<ResourceLocation, KnownTag<T>> tagCache;
+    
+    public KnownTagManager(ResourceKey<? extends Registry<T>> resourceKey, Class<T> elementClass) {
+        
+        this.resourceKey = resourceKey;
+        this.elementClass = elementClass;
+        this.backingResult = new MutableLoadResult<>();
+        this.tagCache = new HashMap<>();
+    }
+    
+    public Optional<Class<?>> elementClass() {
+        
+        return Optional.of(elementClass);
+    }
+    
+    @Override
+    public ResourceKey<? extends Registry<T>> resourceKey() {
+        
+        return resourceKey;
+    }
+    
+    @SafeVarargs
+    @ZenCodeType.Method
+    public final void addElements(KnownTag<T> to, T... values) {
+        
+        if(!exists(to)) {
+            CraftTweakerAPI.apply(new ActionKnownTagCreate<>(to));
+        }
+        CraftTweakerAPI.apply(new ActionKnownTagAdd<>(to, List.of(values)));
+        recalculate();
+    }
+    
+    @ZenCodeType.Method
+    public List<T> elements(KnownTag<T> of) {
+        
+        if(!exists(of)) {
+            return List.of();
+        }
+        return getInternal(of).getValues().stream().map(Holder::value).collect(Collectors.toList());
+    }
+    
+    @SafeVarargs
+    @ZenCodeType.Method
+    public final void removeElements(KnownTag<T> from, T... values) {
+        
+        if(!exists(from)) {
+            throw new IllegalArgumentException("Cannot remove elements from empty tag: " + from);
+        }
+        CraftTweakerAPI.apply(new ActionKnownTagRemove<>(from, List.of(values)));
+        recalculate();
+    }
+    
+    @ZenCodeType.Method
+    public KnownTag<T> tag(String id) {
+        
+        return tag(new ResourceLocation(id));
+    }
+    
+    @ZenCodeType.Method
+    public KnownTag<T> tag(ResourceLocation id) {
```

Comment:
Having a helper for the java side of this that takes a `TagKey` would be of use rather than having to call location manually.